projects/components/src/dropdown/dropdown.component.ts

```diff
@@ -5,24 +5,43 @@
 
 import {
     Component,
+    ElementRef,
+    EventEmitter,
     Host,
     HostListener,
     Input,
     Optional,
+    Output,
+    Provider,
     QueryList,
+    Renderer2,
     SkipSelf,
     TrackByFunction,
     ViewChild,
     ViewChildren,
 } from '@angular/core';
-import { ClrDropdown } from '@clr/angular';
+import { ClrDropdown, ClrDropdownMenu, ClrDropdownTrigger } from '@clr/angular';
 import { TextIcon } from '../common/interfaces';
 import { CliptextConfig, TooltipSize } from '../lib/directives';
+import { DropdownFocusHandlerService } from './dropdown-focus-handler.service';
 
 const NESTED_DROPDOWN_TRIGGER_SELECTOR = 'clr-dropdown clr-dropdown > button';
 const DROPDOWN_ITEM_SELECTOR = 'clr-dropdown-menu > button';
 export const NESTED_MENU_HIDE_DELAY = 400;
 
+export function dropdownFocusHandlerServiceFactory(
+    renderer: Renderer2,
+    existing: DropdownFocusHandlerService
+): DropdownFocusHandlerService {
+    return existing || new DropdownFocusHandlerService(renderer);
+}
+
+export const DROPDOWN_FOCUS_HANDLER_PROVIDER: Provider = {
+    provide: DropdownFocusHandlerService,
+    useFactory: dropdownFocusHandlerServiceFactory,
+    deps: [Renderer2, [new Optional(), new SkipSelf(), DropdownFocusHandlerService]],
```

Comment:
Does this generate a singelton for all dropdowns even if there are many on the same screen? Or is it smart enough to generate a different one per dropdown but not nested dropdowns?

Reply:
It's usually safer to create a function that returns an object. This creates a new copy every time it's called instead of everyone always sharing an object